File: RAnalytics/Core/Private/_RAnalyticsLaunchCollector.m
#import <CommonCrypto/CommonDigest.h>
#import <RAnalytics/RAnalyticsEvent.h>
#import "_RAnalyticsLaunchCollector.h"
#import "_RAnalyticsHelpers.h"
#import <UserNotifications/UserNotifications.h>
#import "RAnalyticsPushTrackingUtility.h"
#import "_RAnalyticsClassManipulator+UNUserNotificationCenter.h"

static NSString *const _RAnalyticsInitialLaunchDateKey = @"com.rakuten.esd.sdk.properties.analytics.launchInformation.initialLaunchDate";
static NSString *const _RAnalyticsInstallLaunchDateKey = @"com.rakuten.esd.sdk.properties.analytics.launchInformation.installLaunchDate";
static NSString *const _RAnalyticsLastUpdateDateKey = @"com.rakuten.esd.sdk.properties.analytics.launchInformation.lastUpdateDate";
static NSString *const _RAnalyticsLastLaunchDateKey = @"com.rakuten.esd.sdk.properties.analytics.launchInformation.lastLaunchDate";
static NSString *const _RAnalyticsLastVersionKey = @"com.rakuten.esd.sdk.properties.analytics.launchInformation.lastVersion";
static NSString *const _RAnalyticsLastVersionLaunchesKey = @"com.rakuten.esd.sdk.properties.analytics.launchInformation.lastVersionLaunches";
static NSTimeInterval const _RAnalyticsPushTapEventTimeLimit = 0.75;

@interface _RAnalyticsLaunchCollector ()
@property (nonatomic, nullable, readwrite, copy) NSDate *initialLaunchDate;
@property (nonatomic, nullable, readwrite, copy) NSDate *installLaunchDate;
@property (nonatomic, nullable, readwrite, copy) NSDate *lastUpdateDate;
@property (nonatomic, nullable, readwrite, copy) NSDate *lastLaunchDate;
@property (nonatomic, nullable, readwrite, copy) NSString *lastVersion;
@property (nonatomic, readwrite) NSUInteger lastVersionLaunches;
@property (nonatomic, readwrite) BOOL isInitialLaunch;
@property (nonatomic, readwrite) BOOL isInstallLaunch;
@property (nonatomic, readwrite) BOOL isUpdateLaunch;
@property (nonatomic, readwrite) RAnalyticsOrigin origin;
@property (nonatomic, nullable, readwrite) UIViewController *currentPage;
@property (nonatomic, nullable, readwrite, copy) NSString *pushTrackingIdentifier;
@property (nonatomic, nullable) NSDate *pushTapTrackingDate;
@end

@implementation _RAnalyticsLaunchCollector

+ (instancetype)sharedInstance
{
    static _RAnalyticsLaunchCollector *instance = nil;
    static dispatch_once_t _RAnalyticsLaunchCollectorOnceToken;
    dispatch_once(&_RAnalyticsLaunchCollectorOnceToken, ^{
        instance = [self.alloc initInstance];
    });
    return instance;
}

- (void)dealloc
{
    [NSNotificationCenter.defaultCenter removeObserver:self];
}

- (instancetype)init
{
    [self doesNotRecognizeSelector:_cmd];
    return nil;
}

- (instancetype)initInstance
{
    if (self = [super init])
    {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(willResume:)
                                                     name:UIApplicationWillEnterForegroundNotification
                                                   object:nil];

        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(didSuspend:)
                                                     name:UIApplicationDidEnterBackgroundNotification
                                                   object:nil];
        
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(didBecomeActive:)
                                                     name:UIApplicationDidBecomeActiveNotification
                                                   object:nil];

        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(didLaunch:)
                                                     name:UIApplicationDidFinishLaunchingNotification
                                                   object:nil];

        // check initLaunchDate exists in keychain
        NSMutableDictionary *query = NSMutableDictionary.new;
        query[(__bridge id)kSecClass]       = (__bridge id)kSecClassGenericPassword;
        query[(__bridge id)kSecAttrLabel]   = _RAnalyticsInitialLaunchDateKey;
        query[(__bridge id)kSecReturnAttributes] = @YES;
        query[(__bridge id)kSecAttrAccessible] = (__bridge id)kSecAttrAccessibleWhenUnlockedThisDeviceOnly;
        CFTypeRef result;
        OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &result);

        if (status == errSecSuccess)
        {
            // keychain item exists
            NSDictionary *values = CFBridgingRelease(result);
            if (!values) return nil;
            _initialLaunchDate = values[(__bridge id)kSecAttrCreationDate];
            _isInitialLaunch = NO;
        }
        else
        {
            // no keychain item
            _initialLaunchDate = NSDate.date;
            query[(__bridge id)kSecAttrCreationDate] = _initialLaunchDate;
            status = SecItemAdd((__bridge CFDictionaryRef)query, 0);
            _isInitialLaunch = YES;
        }

        [self resetToDefaults];
        _isInstallLaunch = (_installLaunchDate) ? NO : YES;
        _isUpdateLaunch = ![_lastVersion isEqualToString:NSBundle.mainBundle.infoDictionary[@"CFBundleShortVersionString"]];
    }
    return self;
}

#pragma mark - App Life Cycle Observers

- (void)willResume:(NSNotification *)notification
{
    [self update];
    [[RAnalyticsEvent.alloc initWithName:RAnalyticsSessionStartEventName parameters:nil] track];
}

- (void)didSuspend:(NSNotification *)notification
{
    [[RAnalyticsEvent.alloc initWithName:RAnalyticsSessionEndEventName parameters:nil] track];
}

- (void)didBecomeActive:(NSNotification *) notification
{
    [self _sendTapNonUNUserNotification];
}

- (void)didLaunch:(NSNotification *) notification
{
    [self update];

    // Equivalent to installation or reinstallation.
    if (_isInitialLaunch)
    {
        [[RAnalyticsEvent.alloc initWithName:RAnalyticsInitialLaunchEventName parameters:nil] track];
        _isInitialLaunch = NO;
    }

    // Triggered on first run after app install with or without version change.
    else if (_isInstallLaunch)
    {
        [[RAnalyticsEvent.alloc initWithName:RAnalyticsInstallEventName parameters:nil] track];
        _isInstallLaunch = NO;
    }

    // Triggered on first run after upgrade (anytime the version number changes).
    else if (_isUpdateLaunch)
    {
        [[RAnalyticsEvent.alloc initWithName:RAnalyticsInstallEventName parameters:nil] track];
        [[RAnalyticsEvent.alloc initWithName:RAnalyticsApplicationUpdateEventName parameters:nil] track];
        _isUpdateLaunch = NO;
    }

    // Trigger a session start.
    [[RAnalyticsEvent.alloc initWithName:RAnalyticsSessionStartEventName parameters:nil] track];

    // Track the credentials status.
    NSDictionary *parameters = @{@"strategies":@{@"password-manager":[self isPasswordExtensionAvailable] ? @"true" : @"false"}};
    [[RAnalyticsEvent.alloc initWithName:RAnalyticsCredentialStrategiesEventName parameters:parameters] track];
}

- (void)didPresentViewController:(UIViewController *)viewController
{
    UIView   *view   = viewController.view;
    UIWindow *window = view.window;

    /*
     * Don't treat as pages view controllers known to be just
     * content-less chromes around other view controllers.
     * Note: won't catch third-party content-less containers.
     */
    if ([viewController isKindOfClass:[UINavigationController class]] ||
        [viewController isKindOfClass:[UISplitViewController class]] ||
        [viewController isKindOfClass:[UIPageViewController class]] ||
        [viewController isKindOfClass:[UITabBarController class]])
    {
        return;
    }
    
    /*
     * Don't treat system popups as pages.
     */
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
    if ([view isKindOfClass:UIAlertView.class] ||
        [view isKindOfClass:UIActionSheet.class] ||
        ([UIAlertController class] && [viewController isKindOfClass:[UIAlertController class]]))
    {
        return;
    }
#pragma clang diagnostic pop

    /*
     * Don't treat private classes as pages if they come from system frameworks.
     * Note: Won't catch private class not adhering to the _ prefix standard.
     */
    if (_RAnalyticsIsApplePrivateClass(viewController.class) ||
        _RAnalyticsIsApplePrivateClass(view.class) ||
        _RAnalyticsIsApplePrivateClass(window.class))
    {
        return;
    }

    /*
     * Allow UIWindow subclasses except those from system frameworks
     * (so that view controllers presented in e.g. UITextEffectWindow are not
     * counting as pages).
     * This catches most keyboard windows.
     */
    if (![window isMemberOfClass:UIWindow.class] && _RAnalyticsIsAppleClass(window.class))
    {
        return;
    }

    /*
     * Keep a strong reference to the view controller in the launch collector only for the
     * time the event is being processed. Note that it will be carried on by the analytics
     * manager state, too.
     */
    _currentPage = viewController;
    [[RAnalyticsEvent.alloc initWithName:RAnalyticsPageVisitEventName parameters:nil] track];
    _currentPage = nil;

    /*
     * Reset the origin to RAnalyticsInternalOrigin for the next page visit after each external
     * call or push notification.
     */
    _origin = RAnalyticsInternalOrigin;
}

#pragma mark - Push Notification

- (void)handleTapNonUNUserNotification: (NSDictionary *) userInfo
                              appState: (UIApplicationState) state
{
    if (_RAnalyticsNotificationsAreHandledByUNDelegate())
    {
        return;
    }
    
    switch (state) {
        case UIApplicationStateBackground:
        case UIApplicationStateInactive:
            _pushTrackingIdentifier = [RAnalyticsPushTrackingUtility trackingIdentifierFromPayload:userInfo];
            _pushTapTrackingDate = [NSDate date];
            break;
        default:
            break;
    }
}

- (void)_sendTapNonUNUserNotification
{
    if (_RAnalyticsNotificationsAreHandledByUNDelegate())
    {
        return;
    }
    
    if (_pushTapTrackingDate != nil &&
        _pushTrackingIdentifier != nil &&
        (fabs(_pushTapTrackingDate.timeIntervalSinceNow) < _RAnalyticsPushTapEventTimeLimit) &&
        ![RAnalyticsPushTrackingUtility analyticsEventHasBeenSentWith:_pushTrackingIdentifier])
    {
        [[RAnalyticsEvent.alloc initWithName:RAnalyticsPushNotificationEventName
                                  parameters:@{RAnalyticsPushNotificationTrackingIdentifierParameter:_pushTrackingIdentifier}] track];
    }
    _pushTrackingIdentifier = nil;
    _pushTapTrackingDate = nil;
}

- (void)processPushNotificationResponse: (UNNotificationResponse*) notificationResponse
{
    UNNotificationTrigger * _Nullable trigger = notificationResponse.notification.request.trigger;
    if (!trigger ||
        ![trigger isKindOfClass:UNPushNotificationTrigger.class]) {
        return;
    }
    
    NSString * _Nullable userText = nil;
    
    if ([trigger isKindOfClass:UNTextInputNotificationResponse.class] &&
        ((UNTextInputNotificationResponse*)trigger).userText.length > 0)
    {
        userText = ((UNTextInputNotificationResponse*)trigger).userText;
    }
    
    [self processPushNotificationPayload:notificationResponse.notification.request.content.userInfo
                              userAction:notificationResponse.actionIdentifier
                                userText:userText];
    
}

- (void)processPushNotificationPayload: (NSDictionary *) userInfo
                            userAction: (NSString *__nullable) userAction
                              userText: (NSString *__nullable) userText
{
    NSString * _Nullable trackingId = [RAnalyticsPushTrackingUtility trackingIdentifierFromPayload:userInfo];
    
    if (trackingId &&
        ![RAnalyticsPushTrackingUtility analyticsEventHasBeenSentWith:trackingId])
    {
        _pushTrackingIdentifier = trackingId;
        [[RAnalyticsEvent.alloc initWithName:RAnalyticsPushNotificationEventName
                                  parameters:@{RAnalyticsPushNotificationTrackingIdentifierParameter:trackingId}] track];
    }
    
    if (_RAnalyticsSharedApplication().applicationState != UIApplicationStateActive)
    {
        // set the origin to push type for the next _rem_visit event
        self.origin = RAnalyticsPushOrigin;
    }
}

- (void)resetToDefaults
{
    NSUserDefaults* defaults = NSUserDefaults.standardUserDefaults;

    id object = [defaults objectForKey:_RAnalyticsInstallLaunchDateKey];
    _installLaunchDate = [object isKindOfClass:NSDate.class] ? object : nil;

    object = [defaults objectForKey:_RAnalyticsLastUpdateDateKey];
    _lastUpdateDate = [object isKindOfClass:NSDate.class] ? object : nil;

    object = [defaults objectForKey:_RAnalyticsLastLaunchDateKey];
    _lastLaunchDate = [object isKindOfClass:NSDate.class] ? object : nil;

    _lastVersion = [defaults stringForKey:_RAnalyticsLastVersionKey];

    object = [defaults objectForKey:_RAnalyticsLastVersionLaunchesKey];
    _lastVersionLaunches = [object isKindOfClass:NSNumber.class] ? [(NSNumber*)object unsignedIntegerValue] : 0;
}

- (void)update
{
    [self resetToDefaults];

    // Update values for the next run
    NSDate *now = NSDate.date;
    NSString *currentVersion = NSBundle.mainBundle.infoDictionary[@"CFBundleShortVersionString"];
    NSUserDefaults* defaults = NSUserDefaults.standardUserDefaults;

    if (!_isInitialLaunch && !_installLaunchDate)
    {
        _isInstallLaunch = YES;
        [defaults setObject:now forKey:_RAnalyticsInstallLaunchDateKey];
    }
    if (![_lastVersion isEqualToString:currentVersion])
    {
        _isUpdateLaunch = YES;
        [defaults setObject:currentVersion forKey:_RAnalyticsLastVersionKey];
        [defaults setObject:now forKey:_RAnalyticsLastUpdateDateKey];
        [defaults setInteger:1 forKey:_RAnalyticsLastVersionLaunchesKey];
    }
    else
    {
        _lastVersionLaunches += 1;
        [defaults setObject:@(_lastVersionLaunches) forKey:_RAnalyticsLastVersionLaunchesKey];
    }
    [defaults setObject:now forKey:_RAnalyticsLastLaunchDateKey];
    [defaults synchronize];
}

- (BOOL)isPasswordExtensionAvailable
{
    if ([NSExtensionItem class]) {
        NSArray*  schemes = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"LSApplicationQueriesSchemes"];
        if( schemes && [schemes isKindOfClass:NSArray.class]) {
            for ( NSString* scheme in schemes) {
                if([scheme isEqualToString:@"org-appextension-feature-password-management"]) {
                    return [_RAnalyticsSharedApplication() canOpenURL:[NSURL URLWithString:@"org-appextension-feature-password-management://"]];
                }
            }
        }
    }
    return NO;
}

@end

File: RAnalytics/Core/Private/RAnalyticsLaunchCollector.swift
import Foundation
import UIKit

enum RAnalyticsLaunchCollectorError: Error {
    case triggerTypeIsIncorrect
    case trackingIsNotProcessed
}

/// This class tracks launch events.
/// It creates event corresponding to each event, sends it to RAnalyticsManager's instance to process.
@objc final class RAnalyticsLaunchCollector: NSObject {
    private enum Constants {
        static let initialLaunchDateKey = "com.rakuten.esd.sdk.properties.analytics.launchInformation.initialLaunchDate"
        static let installLaunchDateKey = "com.rakuten.esd.sdk.properties.analytics.launchInformation.installLaunchDate"
        static let lastUpdateDateKey = "com.rakuten.esd.sdk.properties.analytics.launchInformation.lastUpdateDate"
        static let lastLaunchDateKey = "com.rakuten.esd.sdk.properties.analytics.launchInformation.lastLaunchDate"
        static let lastVersionKey = "com.rakuten.esd.sdk.properties.analytics.launchInformation.lastVersion"
        static let lastVersionLaunchesKey = "com.rakuten.esd.sdk.properties.analytics.launchInformation.lastVersionLaunches"
    }

    /// The initial launch date is being stored in keychain.
    @objc private(set) var initialLaunchDate: Date?

    /// The install launch date is being stored in shared preferences.
    @objc private(set) var installLaunchDate: Date?

    /// The last update date is being stored in shared preferences.
    @objc private(set) var lastUpdateDate: Date?

    /// The last launch date is being stored in shared preferences.
    @objc private(set) var lastLaunchDate: Date?

    /// The last version is being stored in shared preferences.
    @objc private(set) var lastVersion: String?

    /// The number of launches since last version is being stored in shared preferences.
    @objc private(set) var lastVersionLaunches: UInt = 0

    /// String identifying the origin of the launch or visit, if it can be determined.
    /// Default value: .internal
    @objc private(set) var origin: AnalyticsManager.State.Origin = .internal

    /// Currently-visited view controller.
    @objc private(set) var currentPage: UIViewController?

    /// The identifier is computed from push payload.
    /// It is used for tracking push notification. It is also sent together with a push notification event.
    @objc private(set) var pushTrackingIdentifier: String?

    private let notificationHandler: NotificationObservable?
    private let userStorageHandler: UserStorageHandleable?
    private let keychainHandler: KeychainHandleable?
    private let tracker: Trackable?

    private var pushTapTrackingDate: Date?
    private let pushTapEventTimeLimit: TimeInterval = 1.5

    @objc private(set) var isInitialLaunch: Bool = false
    @objc private(set) var isInstallLaunch: Bool = false
    @objc private(set) var isUpdateLaunch: Bool = false

    @available(*, unavailable)
    override init() {
        self.notificationHandler = nil
        self.userStorageHandler = nil
        self.keychainHandler = nil
        self.tracker = nil
        super.init()
    }

    /// Creates a launch collector
    ///
    /// - Parameters:
    ///   - dependenciesFactory: The dependencies factory.
    ///
    /// - Returns: An instance of RAnalyticsLaunchCollector or nil.
    @objc init?(dependenciesFactory: DependenciesFactory) {
        guard let notificationHandler = dependenciesFactory.notificationHandler,
              let userStorageHandler = dependenciesFactory.userStorageHandler,
              let keychainHandler = dependenciesFactory.keychainHandler,
              let tracker = dependenciesFactory.tracker else {
            return nil
        }
        self.notificationHandler = notificationHandler
        self.userStorageHandler = userStorageHandler
        self.keychainHandler = keychainHandler
        self.tracker = tracker

        super.init()

        configureNotifications()

        guard configureLaunchValues() else {
            return nil
        }

        resetToDefaults()
        isInstallLaunch = (installLaunchDate == nil)
        isUpdateLaunch = lastVersion != (Bundle.main.shortVersion ?? "")
    }

    private func configureNotifications() {
        notificationHandler?.addObserver(self,
                                         selector: #selector(willResume(_:)),
                                         name: UIApplication.willEnterForegroundNotification,
                                         object: nil)

        notificationHandler?.addObserver(self,
                                         selector: #selector(didSuspend(_:)),
                                         name: UIApplication.didEnterBackgroundNotification,
                                         object: nil)

        notificationHandler?.addObserver(self,
                                         selector: #selector(didBecomeActive(_:)),
                                         name: UIApplication.didBecomeActiveNotification,
                                         object: nil)

        notificationHandler?.addObserver(self,
                                         selector: #selector(didLaunch(_:)),
                                         name: UIApplication.didFinishLaunchingNotification,
                                         object: nil)
    }

    private func configureLaunchValues() -> Bool {
        // check initLaunchDate exists in keychain
        let response = keychainHandler?.item(for: Constants.initialLaunchDateKey)

        switch response?.status {
        case errSecSuccess:
            // keychain item exists
            guard let date = keychainHandler?.creationDate(for: response?.result) else {
                return false
            }
            initialLaunchDate = date
            isInitialLaunch = false
        default:
            // no keychain item
            initialLaunchDate = Date()
            keychainHandler?.set(creationDate: initialLaunchDate, for: Constants.initialLaunchDateKey)
            isInitialLaunch = true
        }
        return true
    }
}

// MARK: - App Life Cycle Observers

@objc extension RAnalyticsLaunchCollector {
    func willResume(_ notification: NSNotification) {
        update()
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.sessionStart, parameters: nil)
    }

    func didSuspend(_ notification: NSNotification) {
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.sessionEnd, parameters: nil)
    }

    func didBecomeActive(_ notification: NSNotification) {
        sendTapNonUNUserNotification()
    }

    func didLaunch(_ notification: NSNotification) {
        update()

        /// Equivalent to installation or reinstallation.
        if isInitialLaunch {
            tracker?.trackEvent(name: AnalyticsManager.Event.Name.initialLaunch, parameters: nil)
            isInitialLaunch = false
        }

        /// Triggered on first run after app install with or without version change.
        else if isInstallLaunch {
            tracker?.trackEvent(name: AnalyticsManager.Event.Name.install, parameters: nil)
            isInstallLaunch = false
        }

        /// Triggered on first run after upgrade (anytime the version number changes).
        else if isUpdateLaunch {
            tracker?.trackEvent(name: AnalyticsManager.Event.Name.install, parameters: nil)
            tracker?.trackEvent(name: AnalyticsManager.Event.Name.applicationUpdate, parameters: nil)
            isUpdateLaunch = false
        }

        /// Trigger a session start.
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.sessionStart, parameters: nil)

        /// Track the credentials status.
        let parameters = ["strategies": ["password-manager": Bundle.isPasswordExtensionAvailable ? "true" : "false"]]
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.credentialStrategies, parameters: parameters)
    }
}

// MARK: - Presenting View Controller

extension RAnalyticsLaunchCollector {
    /// This method is called when the swizzling method _swizzled_viewDidAppear in _RAnalyticsTrackingPageView is called.
    /// The _swizzled_viewDidAppear is called when the view of an UIViewController is shown.
    func didPresentViewController(_ viewController: UIViewController) {
        guard viewController.isTrackableAsPageVisit else {
            return
        }

        /// Keep a strong reference to the view controller in the launch collector only for the
        /// time the event is being processed. Note that it will be carried on by the analytics
        /// manager state, too.
        currentPage = viewController
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.pageVisit, parameters: nil)
        currentPage = nil

        /// Reset the origin to RAnalyticsInternalOrigin for the next page visit after each external
        /// call or push notification.
        origin = .internal
    }
}

// MARK: - Push Notification

extension RAnalyticsLaunchCollector {
    /// For implementations that do NOT use the UNUserNotification Framework,  We need to distinguish between a tap of notification alert and
    /// receiving a push notification.  This can be done by measuring the time since when this function was called and the next app life cycle "App did become active" event occuring.
    func handleTapNonUNUserNotification(_ userInfo: [AnyHashable: Any], appState state: UIApplication.State) {
        guard !UNUserNotificationCenter.notificationsAreHandledByUNDelegate else {
            return
        }

        switch state {
        case .background, .inactive:
            pushTrackingIdentifier = RAnalyticsPushTrackingUtility.trackingIdentifier(fromPayload: userInfo)
            pushTapTrackingDate = Date()
        default: ()
        }
    }
}

extension RAnalyticsLaunchCollector {
    /// This method sends a push open notify event only if a tracking identifier can be pulled from the UNNotificationResponse
    /// - Returns:
    ///     - success if the trigger is kind of UNPushNotificationTrigger and if the tracking is processed. The associated value is the trigger.
    ///     - or a failure if the trigger is not kind of UNPushNotificationTrigger or if the tracking is not processed.
    @discardableResult
    func processPushNotificationResponse(_ notificationResponse: UNNotificationResponse) -> Result<UNNotificationTrigger, RAnalyticsLaunchCollectorError> {
        guard let trigger = notificationResponse.notification.request.trigger,
              trigger.isKind(of: UNPushNotificationTrigger.self) else {
            return .failure(.triggerTypeIsIncorrect)
        }

        var userText: String?
        if let textInputNotificationResponse = notificationResponse as? UNTextInputNotificationResponse,
           !textInputNotificationResponse.userText.isEmpty {
            userText = textInputNotificationResponse.userText
        }
        let isProcessed = processPushNotificationPayload(userInfo: notificationResponse.notification.request.content.userInfo,
                                                         userAction: notificationResponse.actionIdentifier,
                                                         userText: userText)
        guard isProcessed else {
            return .failure(.trackingIsNotProcessed)
        }
        return .success(trigger)
    }

    /// This method sends a push open notify event only if a tracking identifier can be pulled from the push payload
    /// - Returns: a boolean to know if the push notification event is tracked or not.
    @discardableResult
    func processPushNotificationPayload(userInfo: [AnyHashable: Any], userAction: String?, userText: String?) -> Bool {
        var isProcessed = false
        let trackingId = RAnalyticsPushTrackingUtility.trackingIdentifier(fromPayload: userInfo)

        if let trackingId = trackingId,
           !RAnalyticsPushTrackingUtility.analyticsEventHasBeenSent(with: trackingId) {
            pushTrackingIdentifier = trackingId
            let parameters = [AnalyticsManager.Event.Parameter.pushTrackingIdentifier: trackingId]
            tracker?.trackEvent(name: AnalyticsManager.Event.Name.pushNotification, parameters: parameters)
            isProcessed = true
        }

        if UIApplication.RAnalyticsSharedApplication?.applicationState != .active {
            /// set the origin to push type for the next _rem_visit event
            origin = .push
        }
        return isProcessed
    }
}

private extension RAnalyticsLaunchCollector {
    func sendTapNonUNUserNotification() {
        guard !UNUserNotificationCenter.notificationsAreHandledByUNDelegate else {
            return
        }

        if let pushTapTrackingDate = pushTapTrackingDate,
           let pushTrackingIdentifier = pushTrackingIdentifier,
           fabs(pushTapTrackingDate.timeIntervalSinceNow) < pushTapEventTimeLimit
            && !RAnalyticsPushTrackingUtility.analyticsEventHasBeenSent(with: pushTrackingIdentifier) {
            tracker?.trackEvent(name: AnalyticsManager.Event.Name.pushNotification,
                                parameters: [AnalyticsManager.Event.Parameter.pushTrackingIdentifier: pushTrackingIdentifier])
        }

        pushTrackingIdentifier = nil
        pushTapTrackingDate = nil
    }
}

// MARK: - Utils

extension RAnalyticsLaunchCollector {
    func resetToDefaults() {
        installLaunchDate = userStorageHandler?.object(forKey: Constants.installLaunchDateKey) as? Date
        lastUpdateDate = userStorageHandler?.object(forKey: Constants.lastUpdateDateKey) as? Date
        lastLaunchDate = userStorageHandler?.object(forKey: Constants.lastLaunchDateKey) as? Date
        lastVersion = userStorageHandler?.string(forKey: Constants.lastVersionKey)
        lastVersionLaunches = (userStorageHandler?.object(forKey: Constants.lastVersionLaunchesKey) as? NSNumber)?.uintValue ?? 0
    }

    func resetPushTrackingIdentifier() {
        pushTrackingIdentifier = nil
    }
}

private extension RAnalyticsLaunchCollector {
    func update() {
        resetToDefaults()

        /// Update values for the next run
        let now = Date()
        let currentVersion = Bundle.main.shortVersion

        if !isInitialLaunch && installLaunchDate == nil {
            isInstallLaunch = true
            userStorageHandler?.set(value: now, forKey: Constants.installLaunchDateKey)
        }

        if lastVersion != currentVersion {
            isUpdateLaunch = true
            userStorageHandler?.set(value: currentVersion, forKey: Constants.lastVersionKey)
            userStorageHandler?.set(value: now, forKey: Constants.lastUpdateDateKey)
            userStorageHandler?.set(value: 1, forKey: Constants.lastVersionLaunchesKey)

        } else {
            lastVersionLaunches += 1
            userStorageHandler?.set(value: NSNumber(value: lastVersionLaunches), forKey: Constants.lastVersionLaunchesKey)
        }
        userStorageHandler?.set(value: now, forKey: Constants.lastLaunchDateKey)
    }
}

extension RAnalyticsLaunchCollector {
    func setOrigin(_ value: AnalyticsManager.State.Origin) {
        origin = value
    }
}
